### DISCUSSION TITLE
Loco position system

### DISCUSSION THREAD

**Discussion by CalvinKoung** [bitcraze/discussions#848] [Q&A]
Hi, I want to use loco position system (UWB) to do indoor navigation. Does the website of bitcraze have any toturials or step-by-step instructions to teach how to set the way point let drone to fly. thanks.

**Comment by knmcguire** [comment]
I'm assuming you have already found this tutorial right? https://www.bitcraze.io/documentation/tutorials/getting-started-with-flying-using-lps/

Next steps are to look at the [autonomy examples in the cflib](https://github.com/bitcraze/crazyflie-lib-python/tree/master/examples) or take a look at these step by step guides to work with the cflib library:

- https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/
- https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_motion_commander/
Just mind that the last guide was written for the flowdeck, so just mind that [with this section](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_motion_commander/#step-1-security-before-flying) of that tutorial to point to the right deck identifier.

**Reply by CalvinKoung** [reply]
Yes, those step by step instructions I already complete. And I want to learn how do connect the crazyflie and loco node to communication and write the code to autonomous control the drone in the indoor environment.

**Reply by knmcguire** [reply]
oke great! Especially the last one should already show you how to set setpoints with the python script. Then you are all set to take a look at the autonomy examples I've showed before. There should be no dependency with those with any positioning system so they should work with the loco positioning as well. 

We don't have any other tutorials to show how to work with autonomy but the [/multiranger/multiranger_push.py](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/multiranger/multiranger_push.py) is a good example of a state machine.

**Reply by CalvinKoung** [reply]
OK, I got it. So in the theory I can use the motion commander python instruction to modify the deck I use to change to loco position system ? And I have another two question ?

1. If I haven't client API can I read out the UWB node and deck information ? like distance or connect state. Because I use console.py I can saw the crazyflie receive the node signal, but how can I know the node is normal connect to the deck. 
2. I equipped z-ranger deck and UWB deck in my crazyflie. So if I want to do autonomus fly. I use z-ranger deck to control altitude and then UWB deck to control direction. Is this concept correct ?

Thanks your help.

**Reply by knmcguire** [reply]
1. You can use the [deck.bcloco](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#deckbcloco). I don't know how to check if the nodes are actually connected or not. There are some [logging values](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#tdoa2) that are related to TWR or TDOA that you can look at or look at the source code of the [loco tab of the cfclient](https://github.com/bitcraze/crazyflie-clients-python/blob/master/src/cfclient/ui/tabs/locopositioning_tab.py). Also you can find the documentation about the loco positioning system from [this landing page](https://www.bitcraze.io/documentation/system/positioning/loco-positioning-system/). But before getting into this, if you just make sure that the positioning is done correctly by the cfclient and just do a visual check of the blinking LEDs on the locodeck, then I think that should be fine as a security check. You can also consider adding [this function](https://github.com/bitcraze/crazyflie-lib-python/blob/294e9cb7f6b7d5c7afb0fb1a1147f2201f81779c/examples/autonomy/autonomousSequence.py#L59-L97) to see if the position estimate diverges.
2. Actually not exactly. Both values are being fused in the estimated kalman filter. Check out [this documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/state_estimators/) for some basic explanation. I think that the right 'concept' is to say that the loco positioningi influences the position estimate of xyz and that the z-ranger mostly influences the z position and makes it more accurate in combination with the z-component of the loco positioning.

**Reply by CalvinKoung** [reply]
Ok, thanks your reply. I will try to figure out.

**Reply by CalvinKoung** [reply]
Hi, I'm sorry. I still have some question want to solve. The motion commander toturials use optical flow deck and set the setpoint and default height to change z direction. So if I addind the x, y direction that it will auto communicate with the UWB node ? If I have some misunderstand please let me know. Thanks

**Comment by amitx64** [comment]
@knmcguire , I did not find any `loco deck` code. So, I can look on it how loco deck estimate the position of it own.

**Reply by CalvinKoung** [reply]
I have similiar question.

**Reply by amitx64** [reply]
@CalvinKoung, I just able to find `loco deck` binary.

**Reply by amitx64** [reply]
@CalvinKoung, I am also want to fly drone in an indoor using this UWB localization.

**Reply by CalvinKoung** [reply]
@amitx64 Can you autonomus control the drone use loco position system ? or you are also trying.

**Reply by amitx64** [reply]
@CalvinKoung, I'm also want to do in autonomous control. I ordered the the bundle of bitcraze for that. But, I also have `Decawave DWM1001C-Dev-Kit`.

**Reply by CalvinKoung** [reply]
@amitx64 you want to use other UWB system to do ?

**Reply by amitx64** [reply]
@CalvinKoung, I ordered `Bitcraze` bundle of `loco deck` and `loco node`. But, currently I have with me the `Decawave DWM1001C-Dev-Kit`.

**Reply by CalvinKoung** [reply]
@amitx64 @knmcguire  I followed the instruction to let drone auntonomus fly around the circle. I think the z-ranger deck provide to estimate default height. But the turn direction I didn't know that is z-ranger provide or loco position node to provide.

**Reply by knmcguire** [reply]
@CalvinKoung this is out of scope of this thread. Please make a new discussion if you haven't already

**Reply by CalvinKoung** [reply]
@knmcguire Sorry, I will create a new thread.

**Comment by amitx64** [comment]
> @amitx64 @knmcguire I followed the instruction to let drone auntonomus fly around the circle. I think the z-ranger deck provide to estimate default height. But the turn direction I didn't know that is z-ranger provide or loco position node to provide.

Yeh! I also thinking the same thing Z-ranger provide the height.

**Reply by knmcguire** [reply]
I wouldn't say that the Z-ranger provides the height. It has more influence on the height _estimate_ since it's height _measurement_ more accurate than the height _measurement_ from loco positioning. The crazyflie is able to estimate height on loco posititioning alone without the z-ranger, but the estimate improves with the addition of the z-ranger. This is the principles of the kalman filter and sensor fusion.
